src/main/java/app/linguistai/bmvp/service/gamification/LeaderboardService.java

```diff
@@ -0,0 +1,145 @@
+            if (loggedUser == null) {
+                throw new NotFoundException(User.class.getSimpleName(), true);
+            }
+
+            // Fetch the ranking of the logged-in user among friends
+            IXPRanking loggedUserRanking = xpRepository.findFriendsUserRankByEmail(loggedUser.getId());
+            if (loggedUserRanking == null) {
+                throw new NotFoundException("User's XP ranking among friends", true);
+            }
+            // If page number is not given, then return the page where the user is present
+            if (page == null) {
+                page = getPageNumberByRanking(loggedUserRanking.getRanking(), size);
+            }
+
+            // Create pageable object with the given page number and size for pagination
+            Pageable pageable = PageRequest.of(page, size);
+            // Fetch page of friends' user XP records sorted by experience
+            Page<UserXP> userXPPage = xpRepository.findTopFriendsByExperience(loggedUser.getId(), pageable);
+
+            // Build response object
+            RLeaderboardXP response = buildLeaderboardResponse(loggedUserRanking, userXPPage);
+            log.info("Friends XP leaderboard fetched for user with email {} (Page {}, Size {}). ", email, page, size);
+            return response;
+        } catch (NotFoundException e) {
+            if (e.getObject().equals(User.class.getSimpleName())) {
+                log.error("When fetching friends XP leaderboard, user is not found with email {}", email);
+            } else {
+                log.error("User's XP ranking among friends with email {} not found.", email);
+            }
+            throw e;
+        } catch (Exception e) {
+            log.error("Get friends XP leaderboard failed for email {} (Page {}, Size {}). ", email, page, size, e);
+            throw new SomethingWentWrongException();
+        }
+    }
+
+    // Helper method to build the leaderboard response
+    private RLeaderboardXP buildLeaderboardResponse(IXPRanking loggedUserRanking, Page<UserXP> userXPPage) throws NotFoundException {
+        List<RUserXPRanking> userXPRankings = new ArrayList<>();
+        // Calculate the first user on the given page's rank by "number of previous users + 1" = page no * page size + 1
+        int rank = userXPPage.getNumber() * userXPPage.getSize() + 1;
+
+        // Iterate through user XP records of the given page to build ranking response objects
+        for (UserXP userXP : userXPPage.getContent()) {
+            User user = userXP.getUser();
+
+            RUserXPRanking ranking = RUserXPRanking.builder()
+                    .user(user)
+                    .experience(userXP.getExperience())
+                    .ranking((long) rank++)
+                    .build();
+
+            userXPRankings.add(ranking);
+        }
+        User user = accountRepository.findUserById(loggedUserRanking.getUserIdAsUUID()).orElse(null);
```

Comment:
Also, instead of assigning orElse(null) and then checking and throwing an exception, you can directly do it in one line as: 
```
User user = accountRepository.findUserById(loggedUserRanking.getUserIdAsUUID())
                         .orElseThrow(() -> new NotFoundException(User.class.getSimpleName(), true));
```

Reply:
Yes, changed it - much more readable, thanks!